src/main/java/com/yahoo/bullet/querying/QueryManager.java

```diff
@@ -0,0 +1,252 @@
+ * relevant to your record (after applying any partitioner) using {@link #partition(BulletRecord)}. You can use the
+ * {@link #addQuery(String, Querier)} to add a query to the manager and the {@link #removeAndGetQuery(String)} and
+ * {@link #removeQueries(Set)} methods to remove a query from the manager.
+ */
+@Slf4j
+public class QueryManager {
+    private Map<String, Set<String>> partitioning;
+    private Map<String, Querier> queries;
+    private Partitioner partitioner;
+    private long queriesSeen = 0;
+    private long expectedQueriesSeen = 0;
+
+    public static final int QUANTILE_STEP = 10;
+
+    public enum PartitionStat {
+        QUERY_COUNT, PARTITION_COUNT, ACTUAL_QUERIES_SEEN, EXPECTED_QUERIES_SEEN,
+        STDDEV_PARTITION_SIZE, LARGEST_PARTITION, SMALLEST_PARTITION, DISTRIBUTION
+    }
+
+    // Exposed for testing.
+    static class Partition implements Comparable<Partition> {
+        private final String name;
+        private final int count;
+        static final String DELIMITER = " -> ";
+
+        private Partition(Map.Entry<String, Set<String>> partition) {
+            name = partition.getKey();
+            count = partition.getValue().size();
+        }
+
+        @Override
+        public int compareTo(Partition o) {
+            return count - o.count;
+        }
+
+        @Override
+        public String toString() {
+            return name + DELIMITER + count;
+        }
+    }
+
+    private static class NoPartitioner implements Partitioner {
+        private static final Set<String> EMPTY_KEYS = Collections.singleton("");
+
+        @Override
+        public Set<String> getKeys(Query query) {
+            return EMPTY_KEYS;
+        }
+
+        @Override
+        public Set<String> getKeys(BulletRecord record) {
+            return EMPTY_KEYS;
+        }
+    }
+
+    /**
+     * The constructor that takes a non-null {@link BulletConfig} instance that contains partitioning settings.
+     *
+     * @param config The non-null config.
+     */
+    public QueryManager(BulletConfig config) {
+        boolean enable = config.getAs(BulletConfig.QUERY_PARTITIONER_ENABLE, Boolean.class);
+        if (enable) {
+            partitioner = config.loadConfiguredClass(BulletConfig.QUERY_PARTITIONER_CLASS_NAME);
+            log.info("Partitioning for queries is enabled. Using %s", partitioner.getClass().getName());
+        } else {
+            partitioner = new NoPartitioner();
+        }
+        partitioning = new HashMap<>();
+        queries = new HashMap<>();
+    }
+
+    /**
+     * Adds a configured, initialized query instance {@link Querier} to the manager.
+     *
+     * @param id The query ID.
+     * @param querier A fully initialized (using {@link Querier#initialize()} {@link Querier} instance.
+     */
+    public void addQuery(String id, Querier querier) {
+        Query query = querier.getRunningQuery().getQuery();
```

Comment:
Add a helper function in Querier to getQuery?

Reply:
Sure